tools/halide_image.h

```diff
@@ -319,8 +319,8 @@ class Image {
 
 private:
     template<typename ...Args>
-    T *address_of(int d, int first, const Args &... rest) const {
-        return address_of(d+1, rest...) + buf.stride[d] * (first - buf.min[d]);
+    T *address_of(int d, int first, Args&&... rest) const {
+        return address_of(d+1, std::forward<Args>(rest)...) + buf.stride[d] * (first - buf.min[d]);
```

Comment:
If any of these args actually get passed by reference instead of value (even in a debug build), it would be very bad for performance. I've been fighting related issues with this code where using operator() made tests much much slower because tests are often built with -g -O1. What we really want is to only accept variadic calls to operator() where all the arguments have already been converted to ints. Instead of messing with address_of, would something like address_of(0, first, int(args)...) work below? Not sure of the right syntax for casting everything in a parameter pack to int.